### PR TITLE
FEAT-#4766: Support fsspec URLs in `read_csv` and `read_csv_glob`

### LINKED ISSUE #4766: FEAT: Add support for Distributed loading on GCS buckets
## Context
The company I work for, heavily uses pandas and Google Cloud Storage (GCS) buckets. We were interested in using Modin to improve performance of loading large datasets and improve data processing tasks. 

## Problem
However, it seems as though using GCS bucket urls for the `read_csv()` function of modin, is quite slow. The GCS bucket file patterns are `gs://`. Loading from local memory or S3 buckets works quite well, but the lack of support for GCS buckets results in Modin to default to pandas, which ends up being slower than using the vanilla pandas library!

## Feature Request
We are requesting support for GCS buckets for distributed loading so that we aren't limited to S3 or local file storage only especially since the company I work for, uses Google Suite products quite a lot and does not use Amazon offerings.

### PR DESCRIPTION
**Pull request by pyrito**
Signed-off-by: Karthik Velayutham <vkarthik@ponder.io>

## What do these changes do?

This PR adds proper support for fsspec URLs that matches pandas behavior for `read_csv`. 

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #4766 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date 
- [x] added (Issue Number: PR title (PR Number)) and github username to release notes for next major release

### CODE REVIEW

File: modin/core/io/text/csv_glob_dispatcher.py [hunk 1]
```diff
@@ -331,18 +329,9 @@ def file_exists(cls, file_path: str, storage_options=None) -> bool:
         bool
             True if the path is valid.
         """
-        if isinstance(file_path, str) and is_url(file_path):
```
Comment (RehanSD) [inline modin/core/io/text/csv_glob_dispatcher.py:334]: What if someone passes in a GitHub raw user content file - with an https irl prefix - would that work?

File: modin/experimental/pandas/test/test_io_exp.py [hunk 2]
```diff
@@ -168,7 +168,7 @@ def test_read_single_csv_with_parse_dates(self, parse_dates):
     Engine.get() != "Ray", reason="Currently only support Ray engine for glob paths."
 )
 def test_read_multiple_csv_s3():
-    path = "S3://modin-datasets/testing/multiple_csv/test_data*.csv"
+    path = "s3://modin-datasets/testing/multiple_csv/test_data*.csv"
```
Comment (RehanSD) [inline modin/experimental/pandas/test/test_io_exp.py:171]: Why was this changed?

File: modin/pandas/test/test_io.py [hunk 3]
```diff
@@ -915,6 +915,13 @@ def test_read_csv_categories(self):
             dtype={"one": "int64", "two": "category"},
         )
```
Comment (RehanSD) [inline modin/pandas/test/test_io.py:917]: Shouldn't we add a test for glob as well?

**Review by RehanSD — COMMENTED** [review]
Left some questions - great work!

File: modin/pandas/test/test_io.py (hunk 3 above)
Comment (pyrito) [reply modin/pandas/test/test_io.py:917]: We already have some glob tests for S3 that should test the new code paths but I can add one for gs as well.

File: modin/experimental/pandas/test/test_io_exp.py (hunk 2 above)
Comment (pyrito) [reply modin/experimental/pandas/test/test_io_exp.py:171]: It is simpler for us to not deal with malformed protocols in the paths and this is what `fsspec` decided to do.

File: modin/core/io/text/csv_glob_dispatcher.py (hunk 1 above)
Comment (pyrito) [reply modin/core/io/text/csv_glob_dispatcher.py:334]: Yeah actually I think it shouldn't work for the glob case. I remember Alexey bringing this up, so I'll revert this back.

File: modin/experimental/pandas/test/test_io_exp.py (hunk 2 above)
Comment (pyrito) [reply modin/experimental/pandas/test/test_io_exp.py]: For more context: https://github.com/fsspec/filesystem_spec/issues/1009

File: modin/pandas/test/test_io.py [hunk 4]
```diff
@@ -915,6 +915,13 @@ def test_read_csv_categories(self):
...
 
+    def test_read_csv_gs(self):
```
Comment (mvashishtha) [inline modin/pandas/test/test_io.py:918]:
```suggestion
    def test_read_csv_google_cloud_storage(self):
```

File: modin/experimental/pandas/test/test_io_exp.py [hunk 5]
```diff
@@ -167,8 +167,11 @@ def test_read_single_csv_with_parse_dates(self, parse_dates):
 @pytest.mark.skipif(
     Engine.get() != "Ray", reason="Currently only support Ray engine for glob paths."
 )
-def test_read_multiple_csv_s3():
-    path = "S3://modin-datasets/testing/multiple_csv/test_data*.csv"
+def test_read_multiple_csv_cloud_store():
+    paths = [
```
Comment (mvashishtha) [inline modin/experimental/pandas/test/test_io_exp.py:171]: Can make these test parameters?

**Review by mvashishtha — COMMENTED** [review]
Overall LGTM. I have two minor comments.

**Review by mvashishtha — DISMISSED** [review]
LGTM!

File: modin/core/io/text/text_file_dispatcher.py [hunk 6]
```diff
@@ -963,6 +964,7 @@ def _read(cls, filepath_or_buffer, **kwargs):
         new_query_compiler : BaseQueryCompiler
             Query compiler with imported data for further processing.
         """
+        filepath_or_buffer = stringify_path(filepath_or_buffer)
```
Comment (anmyachev) [inline modin/core/io/text/text_file_dispatcher.py:967]: Why is this needed?

Comment (pyrito) [reply modin/core/io/text/text_file_dispatcher.py:967]: @anmyachev it's a nice utility function used by pandas. Sometimes users pass in Path-like objects and we need to attempt to stringify the path. If it ends up being just a buffer, this method will handle that for us as well.

Comment (anmyachev) [reply modin/core/io/text/text_file_dispatcher.py:967]: Doesn't this method do part of the job as our functions `get_path` and `get_path_or_buffer`? In this case, we could refuse to use our functions.

Comment (pyrito) [reply modin/core/io/text/text_file_dispatcher.py:967]: Not for get_path since that function expects a string to be passed in. Unsure about the latter

Comment (anmyachev) [reply modin/core/io/text/text_file_dispatcher.py:967]: It seems to me the code before worked with Path-like types without `stringify_path`, didn't it?

Comment (pyrito) [reply modin/core/io/text/text_file_dispatcher.py:967]: Yes but now that I think about it, it might be better to use stringify_path in place of both get_path and get_path_or_buffer.

Comment (pyrito) [reply modin/core/io/text/text_file_dispatcher.py:967]: @anmyachev reading the warning in `get_path_or_buffer` makes me wonder if we should leave our current implementation as is?

Comment (anmyachev) [reply modin/core/io/text/text_file_dispatcher.py:967]: Yes, `stringify_path` cannot replace `get_path_or_buffer`.

Comment (anmyachev) [reply modin/core/io/text/text_file_dispatcher.py:967]: We can try to replace `get_path` with `stringify_path` in a separate pull request if needed.

Comment (pyrito) [reply modin/core/io/text/text_file_dispatcher.py]: @anmyachev yes, I reverted things back for now. Please take a look and approve if it looks good!

File: modin/core/io/text/csv_glob_dispatcher.py [hunk 7]
```diff
@@ -54,7 +52,7 @@ def _read(cls, filepath_or_buffer, **kwargs):
             Query compiler with imported data for further processing.
         """
         # Ensures that the file is a string file path. Otherwise, default to pandas.
-        filepath_or_buffer = cls.get_path_or_buffer(filepath_or_buffer)
+        filepath_or_buffer = cls.get_path_or_buffer(stringify_path(filepath_or_buffer))
```
Comment (anmyachev) [inline modin/core/io/text/csv_glob_dispatcher.py:55]: `stringify_path` is needed here because `get_path` isn't called?

Comment (pyrito) [reply modin/core/io/text/csv_glob_dispatcher.py:55]: `get_path` in the glob file reader returns a list of files, so `stringify_path` makes sense here.

**Review by mvashishtha — APPROVED** [review]

**Review by anmyachev — APPROVED** [review]
